File: python/tests/test_transport_image_processed.py
"""v0.11 Task 5.1 (Deferral 1, transport-spec §4.17 processed mode):
exercise sparse :data:`PacketType.IMAGE_PIXEL` (0x14) emission on
:meth:`TransportWriter.write_image_processed` and the matching decode
path on :class:`TransportReader`.

Processed mode emits, per pixel, a list of
``(channel_index, intensity)`` pairs indexed into the shared
``mz_axis`` (carried on the IMAGE_HEADER). The MSImage data model is
unchanged - the dense ``intensity`` cube round-trips byte-for-byte
through the sparse wire shape. Channels that are not enumerated in
the payload decode as 0.0.

Wire layout (LITTLE-ENDIAN, see transport-spec §4.17)::

    x(u32), y(u32), precision(u8), compression(u8),
    payload_length(u32),
    payload_bytes = nonzero_count(u32)
                  + nonzero_count x { channel_index(u32) + intensity(fXX) }

``write_image`` (continuous mode) is unchanged and remains the default
emission path; ``write_image_processed`` is an opt-in sibling for
sparse cubes. Java parity:
``TransportImageProcessedTest`` (commit ``1889343e``).

SPDX-License-Identifier: Apache-2.0
"""
from __future__ import annotations

import io
import struct
from pathlib import Path

import numpy as np

from ttio import MSImage
from ttio.transport.codec import TransportReader, TransportWriter
from ttio.transport.packets import PacketType


def _build_sparse_image() -> MSImage:
    """3x3x10 cube, ~80% sparse - per-pixel nonzeros < bins/4.

    Mirrors Java's ``buildSparseImage`` so the on-wire byte layout
    matches across languages: each pixel gets 1-2 nonzero channels
    deterministically.
    """
    w, h, bins = 3, 3, 10
    cube = np.zeros((h, w, bins), dtype=np.float64)
    for y in range(h):
        for x in range(w):
            p = y * w + x
            cube[y, x, p % bins] = 1.0 + p
            if (p & 1) == 1:
                cube[y, x, (p * 3) % bins] = -2.0 - p
    mz = np.array([100.0 + k * 5.0 for k in range(bins)], dtype=np.float64)
    return MSImage(
        width=w, height=h, spectral_points=bins,
        intensity=cube, mz_axis=mz,
        pixel_size_x=10.0, pixel_size_y=10.0, scan_pattern="raster",
        title="sparse", isa_investigation_id="",
    )


def _encode_image_processed(image: MSImage) -> bytes:
    """Encode an MSImage via writeImageProcessed only (no stream framing)."""
    sink = io.BytesIO()
    with TransportWriter(sink) as w:
        w.write_image_processed(image)
    return sink.getvalue()


def _decode_to_image(data: bytes) -> MSImage:
    """Parse the IMAGE_HEADER + IMAGE_PIXEL records directly into a
    dense MSImage. Mirrors Java's ``decodeToImage`` helper - we don't
    go through ``read_to_dataset`` because the test only writes an
    image block (no StreamHeader)."""
    r = TransportReader(io.BytesIO(data))
    records = r.records_for_test()
    hdr_idx = -1
    for i, rec in enumerate(records):
        if rec.header.packet_type == int(PacketType.IMAGE_HEADER):
            hdr_idx = i
            break
    assert hdr_idx >= 0, "must contain an IMAGE_HEADER"

    hdr = records[hdr_idx].payload
    off = 0
    off += 1                                                # modality
    (width,) = struct.unpack_from("<I", hdr, off); off += 4
    (height,) = struct.unpack_from("<I", hdr, off); off += 4
    (bins,) = struct.unpack_from("<I", hdr, off); off += 4
    (px_x,) = struct.unpack_from("<d", hdr, off); off += 8
    (px_y,) = struct.unpack_from("<d", hdr, off); off += 8
    off += 1                                                # scan_pattern
    off += 1                                                # axis_kind
    (axis_len,) = struct.unpack_from("<I", hdr, off); off += 4
    mz = np.frombuffer(hdr, dtype="<f8", count=axis_len, offset=off).copy()
    off += 8 * axis_len
    is_cont = hdr[off]; off += 1
    assert is_cont == 0, "decode_to_image expects processed-mode header"

    cube = np.zeros((height, width, bins), dtype=np.float64)
    for rec in records[hdr_idx + 1:]:
        if rec.header.packet_type == int(PacketType.END_OF_IMAGE):
            break
        if rec.header.packet_type != int(PacketType.IMAGE_PIXEL):
            continue
        pl = rec.payload
        off = 0
        (x,) = struct.unpack_from("<I", pl, off); off += 4
        (y,) = struct.unpack_from("<I", pl, off); off += 4
        precision = pl[off]; off += 1
        compression = pl[off]; off += 1
        (plen,) = struct.unpack_from("<I", pl, off); off += 4
        raw = bytes(pl[off:off + plen])
        assert compression == 0, (
            "compression=0 is the only mode produced today"
        )
        (count,) = struct.unpack_from("<I", raw, 0)
        eo = 4
        for _ in range(count):
            (ch,) = struct.unpack_from("<I", raw, eo); eo += 4
            if precision == 1:
                (v,) = struct.unpack_from("<d", raw, eo); eo += 8
            else:
                (v,) = struct.unpack_from("<f", raw, eo); eo += 4
            cube[y, x, ch] = v

    return MSImage(
        width=width, height=height, spectral_points=bins,
        intensity=cube, mz_axis=mz,
        pixel_size_x=px_x, pixel_size_y=px_y, scan_pattern="raster",
    )


def test_processed_mode_round_trips_sparse_cube() -> None:
    """Round-trip a sparse 3x3x10 MSImage through processed-mode
    encode -> decode and assert the dense intensity cube matches the
    input byte-for-byte. The fixture is ~80% sparse: each pixel has a
    small handful of nonzero channels seeded by a deterministic
    pattern."""
    img = _build_sparse_image()
    data = _encode_image_processed(img)

    r = TransportReader(io.BytesIO(data))
    records = r.records_for_test()
    # 1 IMAGE_HEADER + 9 IMAGE_PIXEL + 1 END_OF_IMAGE
    assert len(records) == 11, (
        f"expected 1 header + 9 pixels + 1 EOI, got {len(records)}"
    )
    assert records[0].header.packet_type == int(PacketType.IMAGE_HEADER)
    for i in range(9):
        assert records[1 + i].header.packet_type == int(PacketType.IMAGE_PIXEL), (
            f"packet {1 + i} must be IMAGE_PIXEL"
        )
    assert records[10].header.packet_type == int(PacketType.END_OF_IMAGE)

    # IMAGE_HEADER must advertise is_continuous=0.
    hdr = records[0].payload
    off = 0
    off += 1                                            # modality
    off += 4 + 4 + 4                                    # width, height, bins
    off += 8 + 8                                        # pixel_size_x/y
    off += 1 + 1                                        # scan_pattern, axis_kind
    (axis_len,) = struct.unpack_from("<I", hdr, off); off += 4
    off += 8 * axis_len                                 # axis values
    is_cont = hdr[off]
    assert is_cont == 0, (
        "write_image_processed must emit is_continuous=0"
    )

    # Materialise into a fresh MSImage and compare cube byte-for-byte.
    rt = _decode_to_image(data)
    assert rt.width == img.width
    assert rt.height == img.height
    assert rt.spectral_points == img.spectral_points
    np.testing.assert_array_equal(rt.mz_axis, img.mz_axis)
    np.testing.assert_array_equal(rt.intensity, img.intensity)


def test_all_zero_pixel_emits_nonzero_count_zero() -> None:
    """An entire pixel with all-zero intensities must encode as
    ``nonzero_count=0``. After round-tripping, the corresponding cube
    slice stays at 0.0 and the wire payload is tiny (just the outer
    header + the 4-byte count)."""
    w, h, bins = 2, 1, 5
    cube = np.zeros((h, w, bins), dtype=np.float64)  # all zero
    mz = np.array([100.0, 110.0, 120.0, 130.0, 140.0], dtype=np.float64)
    img = MSImage(
        width=w, height=h, spectral_points=bins,
        intensity=cube, mz_axis=mz,
        pixel_size_x=10.0, pixel_size_y=10.0, scan_pattern="raster",
        title="all_zero", isa_investigation_id="",
    )

    data = _encode_image_processed(img)
    r = TransportReader(io.BytesIO(data))
    records = r.records_for_test()
    # 1 hdr + 2 pixels + 1 EOI
    assert len(records) == 4

    # Each IMAGE_PIXEL payload after the outer fixed header must carry
    # nonzero_count=0 (a single 4-byte u32 LE = 0).
    for i in range(2):
        pl = records[1 + i].payload
        off = 0
        off += 4                                       # x
        off += 4                                       # y
        off += 1 + 1                                   # precision, compression
        (plen,) = struct.unpack_from("<I", pl, off); off += 4
        assert plen == 4, (
            "all-zero pixel must have payload_length=4 (just nonzero_count)"
        )
        (count,) = struct.unpack_from("<I", pl, off)
        assert count == 0, "all-zero pixel must emit nonzero_count=0"

    rt = _decode_to_image(data)
    np.testing.assert_array_equal(rt.intensity, cube)


def test_fully_dense_pixel_round_trips_in_processed_mode() -> None:
    """A fully dense pixel (every channel nonzero) round-trips
    correctly even in processed mode - no off-by-one on the
    channel-index loop."""
    w, h, bins = 1, 1, 8
    cube = np.zeros((h, w, bins), dtype=np.float64)
    for k in range(bins):
        cube[0, 0, k] = (k + 1) * 7.5  # all nonzero
    mz = np.array([200.0 + k for k in range(bins)], dtype=np.float64)
    img = MSImage(
        width=w, height=h, spectral_points=bins,
        intensity=cube, mz_axis=mz,
        pixel_size_x=5.0, pixel_size_y=5.0, scan_pattern="raster",
        title="dense", isa_investigation_id="",
    )

    data = _encode_image_processed(img)
    r = TransportReader(io.BytesIO(data))
    records = r.records_for_test()
    # 1 hdr + 1 pixel + 1 EOI
    assert len(records) == 3

    # Inspect the pixel payload: nonzero_count must be bins, and
    # payload_length = 4 + bins*(4+8).
    pl = records[1].payload
    off = 4 + 4 + 1 + 1                                # x, y, prec, comp
    (plen,) = struct.unpack_from("<I", pl, off); off += 4
    assert plen == 4 + bins * (4 + 8), (
        "dense processed-mode pixel payload size mismatch"
    )
    (count,) = struct.unpack_from("<I", pl, off); off += 4
    assert count == bins
    for k in range(bins):
        (ch,) = struct.unpack_from("<I", pl, off); off += 4
        (v,) = struct.unpack_from("<d", pl, off); off += 8
        assert ch == k, (
            "channels must arrive in ascending order for a dense pixel"
        )
        assert v == (k + 1) * 7.5

    rt = _decode_to_image(data)
    np.testing.assert_array_equal(rt.intensity, cube)


def test_mixed_sparse_and_dense_pixels_round_trip() -> None:
    """Mixed sparse-and-dense pixels in the same image round-trip
    correctly: row 0 is dense, row 1 is sparse, row 2 is empty. All
    three coexist on the wire under a single IMAGE_HEADER."""
    w, h, bins = 2, 3, 6
    cube = np.zeros((h, w, bins), dtype=np.float64)
    # Row 0: dense.
    for x in range(w):
        for k in range(bins):
            cube[0, x, k] = (k + 1) + x * 0.5
    # Row 1: sparse - only channels 0, 3 are nonzero.
    for x in range(w):
        cube[1, x, 0] = 99.0 + x
        cube[1, x, 3] = -1.5 * (x + 1)
    # Row 2: empty (all zeros - left as default).
    mz = np.array([300.0 + k for k in range(bins)], dtype=np.float64)
    img = MSImage(
        width=w, height=h, spectral_points=bins,
        intensity=cube, mz_axis=mz,
        pixel_size_x=1.0, pixel_size_y=1.0, scan_pattern="raster",
        title="mixed", isa_investigation_id="",
    )

    data = _encode_image_processed(img)
    rt = _decode_to_image(data)
    np.testing.assert_array_equal(
        rt.intensity, cube,
        err_msg="mixed-sparsity cube must round-trip exactly",
    )
    np.testing.assert_array_equal(rt.mz_axis, mz)

File: objc/Tests/TestTransportImageProcessed.m
/*
 * TestTransportImageProcessed.m — Task 5.1 of transport-spec v0.11
 * (Deferral 1).
 *
 * Exercises -[TTIOTransportWriter writeImageProcessed:] (sparse
 * IMAGE_PIXEL payloads) + the matching reader path that reconstructs
 * the dense intensity cube. The TTIOMSImage data model stays dense;
 * processed mode is purely a wire optimisation for sparse cubes.
 *
 * Wire layout per transport-spec §4.17 (LITTLE-ENDIAN). The
 * IMAGE_HEADER is identical to -writeImage: except for
 * is_continuous == 0; each IMAGE_PIXEL payload is:
 *
 *   x(u32) + y(u32) + precision(u8) + compression(u8) +
 *   payload_length(u32) +
 *   payload_bytes = nonzero_count(u32) +
 *     nonzero_count × { channel_index(u32) + intensity(f64) }
 *
 * Cross-language parity:
 *   Java TransportImageProcessedTest (commit 1889343e)
 *   Python tests/test_transport_image_processed.py (commit 8eac605a)
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"

#import "Transport/TTIOTransportPacket.h"
#import "Transport/TTIOTransportReader.h"
#import "Transport/TTIOTransportReader+Internal.h"
#import "Transport/TTIOTransportWriter.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Image/TTIOMSImage.h"
#include <unistd.h>

// -------- helpers ----------------------------------------------------------

static NSString *makeTempPathIp(NSString *suffix)
{
    NSString *base = [NSString stringWithFormat:@"ttio_tr_imgp_%d_%@.tio",
                       (int)getpid(), suffix];
    return [NSTemporaryDirectory() stringByAppendingPathComponent:base];
}

static uint32_t leU32Ip(const uint8_t *b)
{
    return (uint32_t)b[0]
         | ((uint32_t)b[1] << 8)
         | ((uint32_t)b[2] << 16)
         | ((uint32_t)b[3] << 24);
}

static double leF64Ip(const uint8_t *b)
{
    uint64_t lo = (uint64_t)leU32Ip(b);
    uint64_t hi = (uint64_t)leU32Ip(b + 4);
    uint64_t bits = lo | (hi << 32);
    double d;
    memcpy(&d, &bits, 8);
    return d;
}

// Sparse 3x3x10 cube: each pixel has a deterministic handful of
// nonzero channels (mirrors Python _build_sparse_image fixture).
static TTIOMSImage *buildSparseFixture(void)
{
    const NSUInteger w = 3;
    const NSUInteger h = 3;
    const NSUInteger s = 10;
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * s * sizeof(double)];
    double *p = (double *)cube.mutableBytes;
    // Initialise to zero, then seed sparse nonzeros.
    for (NSUInteger y = 0; y < h; y++) {
        for (NSUInteger x = 0; x < w; x++) {
            NSUInteger pixelIdx = x + y * w;
            NSUInteger base = (y * w + x) * s;
            // 2 nonzero channels per pixel at varying positions.
            NSUInteger ch1 = pixelIdx % s;
            NSUInteger ch2 = (pixelIdx * 3 + 7) % s;
            p[base + ch1] = (double)(pixelIdx + 1) * 1.5;
            if (ch2 != ch1) {
                p[base + ch2] = (double)(pixelIdx + 1) * 2.5;
            }
        }
    }
    NSMutableData *mz = [NSMutableData dataWithLength:s * sizeof(double)];
    double *mzp = (double *)mz.mutableBytes;
    for (NSUInteger i = 0; i < s; i++) mzp[i] = 100.0 + (double)i * 10.0;

    return [[TTIOMSImage alloc]
                initWithTitle:@"sparse_processed"
           isaInvestigationId:@""
              identifications:@[]
              quantifications:@[]
            provenanceRecords:@[]
                        width:w
                       height:h
               spectralPoints:s
                     tileSize:32
                   pixelSizeX:10.0
                   pixelSizeY:10.0
                  scanPattern:@"raster"
                         cube:cube
                       mzAxis:mz];
}

// -------- 1. round-trip a sparse cube --------------------------------------

static void testProcessedModeRoundTripsSparseCube(void)
{
    TTIOMSImage *img = buildSparseFixture();

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"img-proc-rt"
                               isaInvestigation:@""
                                       features:@[TTIOTransportV011Feature]
                                      nDatasets:0
                                          error:&err],
         "5.1 rt: StreamHeader");
    PASS([w writeImageProcessed:img error:&err],
         "5.1 rt: writeImageProcessed: emitted");
    PASS([w writeEndOfStreamWithError:&err], "5.1 rt: EndOfStream");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    // StreamHeader + IMAGE_HEADER + 9 IMAGE_PIXEL + END_OF_IMAGE + EndOfStream
    PASS(records.count == 1 + 1 + 9 + 1 + 1,
         "5.1 rt: 13 packets (header + 1 + 9 + 1 + EOS)");

    PASS(records[1].header.packetType
            == TTIOTransportPacketImageHeader,
         "5.1 rt: packet 1 is IMAGE_HEADER (0x13)");

    // Verify is_continuous=0 on the wire.
    NSData *hdrPayload = records[1].payload;
    const uint8_t *hb = hdrPayload.bytes;
    // is_continuous offset: 1 (modality) + 4 + 4 + 4 (w,h,bins) +
    // 8 + 8 (pxX, pxY) + 1 (scan) + 1 (axis_kind) + 4 (axis_length)
    // + 8 * axis_length(10) = 113
    NSUInteger isContOff = 1 + 4 + 4 + 4 + 8 + 8 + 1 + 1 + 4 + 8 * 10;
    PASS(hb[isContOff] == 0u,
         "5.1 rt: IMAGE_HEADER is_continuous == 0 (processed mode)");

    // Materialise the stream back into a .tio and re-read the MSImage.
    NSString *outPath = makeTempPathIp(@"rt");
    unlink([outPath fileSystemRepresentation]);
    TTIOTransportReader *r2 = [[TTIOTransportReader alloc] initWithData:buf];
    PASS([r2 writeTtioToPath:outPath error:&err] && err == nil,
         "5.1 rt: writeTtioToPath materialised processed-mode stream");

    TTIOMSImage *back = [TTIOMSImage readFromFilePath:outPath error:&err];
    PASS(back != nil && err == nil,
         "5.1 rt: re-opened materialised image");
    PASS(back.width == img.width
        && back.height == img.height
        && back.spectralPoints == img.spectralPoints,
         "5.1 rt: round-trip dims match");
    PASS([back.cube isEqualToData:img.cube],
         "5.1 rt: round-trip cube byte-equal (sparse -> dense)");
    PASS([back.mzAxis isEqualToData:img.mzAxis],
         "5.1 rt: round-trip mz_axis byte-equal");
    unlink([outPath fileSystemRepresentation]);
}

// -------- 2. all-zero pixel emits nonzero_count = 0 ------------------------

static void testAllZeroPixelEmitsNonzeroCountZero(void)
{
    const NSUInteger w = 2, h = 1, s = 5;
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * s * sizeof(double)];
    // intentionally left zero
    NSMutableData *mz = [NSMutableData dataWithLength:s * sizeof(double)];
    double *mzp = (double *)mz.mutableBytes;
    for (NSUInteger i = 0; i < s; i++) mzp[i] = 100.0 + (double)i * 10.0;
    TTIOMSImage *img = [[TTIOMSImage alloc]
                            initWithTitle:@"all_zero"
                       isaInvestigationId:@""
                          identifications:@[]
                          quantifications:@[]
                        provenanceRecords:@[]
                                    width:w
                                   height:h
                           spectralPoints:s
                                 tileSize:32
                               pixelSizeX:10.0
                               pixelSizeY:10.0
                              scanPattern:@"raster"
                                     cube:cube
                                   mzAxis:mz];

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w2 =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w2 writeStreamHeaderWithFormatVersion:@"1.2"
                                           title:@"img-zero"
                                isaInvestigation:@""
                                        features:@[]
                                       nDatasets:0
                                           error:&err],
         "5.1 zero: StreamHeader");
    PASS([w2 writeImageProcessed:img error:&err],
         "5.1 zero: writeImageProcessed");
    PASS([w2 writeEndOfStreamWithError:&err], "5.1 zero: EndOfStream");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    // SH + IMAGE_HEADER + 2 IMAGE_PIXEL + EOI + EOS = 6
    PASS(records.count == 6,
         "5.1 zero: 6 packets total");

    // Each IMAGE_PIXEL: payload_length must be exactly 4 (the
    // nonzero_count field alone), and nonzero_count must be 0.
    for (NSUInteger i = 0; i < 2; i++) {
        TTIOTransportPacketRecord *rec = records[2 + i];
        const uint8_t *pb = rec.payload.bytes;
        NSUInteger po = 0;
        po += 4;       // x
        po += 4;       // y
        po += 1 + 1;   // precision + compression
        uint32_t payloadLen = leU32Ip(&pb[po]); po += 4;
        PASS(payloadLen == 4u,
             "5.1 zero: all-zero pixel payload_length == 4 (just nonzero_count)");
        uint32_t nonzero = leU32Ip(&pb[po]);
        PASS(nonzero == 0u,
             "5.1 zero: all-zero pixel nonzero_count == 0");
    }
}

// -------- 3. fully-dense pixel round-trips correctly ------------------------

static void testFullyDensePixelRoundTrips(void)
{
    const NSUInteger w = 1, h = 1, s = 8;
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * s * sizeof(double)];
    double *p = (double *)cube.mutableBytes;
    for (NSUInteger k = 0; k < s; k++) p[k] = (double)(k + 1) * 7.5;
    NSMutableData *mz = [NSMutableData dataWithLength:s * sizeof(double)];
    double *mzp = (double *)mz.mutableBytes;
    for (NSUInteger i = 0; i < s; i++) mzp[i] = 200.0 + (double)i;
    TTIOMSImage *img = [[TTIOMSImage alloc]
                            initWithTitle:@"dense"
                       isaInvestigationId:@""
                          identifications:@[]
                          quantifications:@[]
                        provenanceRecords:@[]
                                    width:w
                                   height:h
                           spectralPoints:s
                                 tileSize:32
                               pixelSizeX:5.0
                               pixelSizeY:5.0
                              scanPattern:@"raster"
                                     cube:cube
                                   mzAxis:mz];

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w2 =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w2 writeStreamHeaderWithFormatVersion:@"1.2"
                                           title:@"img-dense"
                                isaInvestigation:@""
                                        features:@[TTIOTransportV011Feature]
                                       nDatasets:0
                                           error:&err],
         "5.1 dense: StreamHeader");
    PASS([w2 writeImageProcessed:img error:&err],
         "5.1 dense: writeImageProcessed");
    PASS([w2 writeEndOfStreamWithError:&err], "5.1 dense: EndOfStream");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records.count == 5,
         "5.1 dense: SH + IMAGE_HEADER + 1 IMAGE_PIXEL + EOI + EOS");

    TTIOTransportPacketRecord *pixRec = records[2];
    const uint8_t *pb = pixRec.payload.bytes;
    NSUInteger po = 4 + 4 + 1 + 1;  // skip x, y, precision, compression
    uint32_t payloadLen = leU32Ip(&pb[po]); po += 4;
    // Fully-dense pixel: payload = 4 + s*(4+8) = 4 + 8*12 = 100.
    PASS(payloadLen == (uint32_t)(4 + s * (4 + 8)),
         "5.1 dense: fully-dense pixel payload size");
    uint32_t nonzero = leU32Ip(&pb[po]); po += 4;
    PASS(nonzero == (uint32_t)s,
         "5.1 dense: nonzero_count == bins (every channel)");
    // Each entry: u32 channel + f64 intensity. Channels MUST appear
    // in ascending order (matches Java/Python iteration).
    for (NSUInteger k = 0; k < s; k++) {
        uint32_t ch = leU32Ip(&pb[po]); po += 4;
        double v = leF64Ip(&pb[po]);    po += 8;
        PASS(ch == (uint32_t)k,
             "5.1 dense: channels ascending");
        PASS(v == (double)(k + 1) * 7.5,
             "5.1 dense: intensities round-trip byte-exact");
    }

    // Full round-trip via materialise.
    NSString *outPath = makeTempPathIp(@"dense");
    unlink([outPath fileSystemRepresentation]);
    TTIOTransportReader *r2 = [[TTIOTransportReader alloc] initWithData:buf];
    PASS([r2 writeTtioToPath:outPath error:&err] && err == nil,
         "5.1 dense: writeTtioToPath materialised");
    TTIOMSImage *back = [TTIOMSImage readFromFilePath:outPath error:&err];
    PASS([back.cube isEqualToData:img.cube],
         "5.1 dense: round-trip cube byte-equal");
    unlink([outPath fileSystemRepresentation]);
}

// -------- 4. mixed sparse + dense pixels round-trip ------------------------

static void testMixedSparseAndDensePixels(void)
{
    const NSUInteger w = 2, h = 2, s = 4;
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * s * sizeof(double)];
    double *p = (double *)cube.mutableBytes;
    // pixel (0,0) all zero; (1,0) one nonzero; (0,1) all nonzero;
    // (1,1) two nonzeros at non-contiguous channels.
    p[(0 * w + 1) * s + 2] = 7.25;
    for (NSUInteger k = 0; k < s; k++) p[(1 * w + 0) * s + k] = (double)(k + 1) * 3.5;
    p[(1 * w + 1) * s + 0] = 11.0;
    p[(1 * w + 1) * s + 3] = 99.5;
    NSMutableData *mz = [NSMutableData dataWithLength:s * sizeof(double)];
    double *mzp = (double *)mz.mutableBytes;
    for (NSUInteger i = 0; i < s; i++) mzp[i] = 50.0 + (double)i * 5.0;

    TTIOMSImage *img = [[TTIOMSImage alloc]
                            initWithTitle:@"mixed"
                       isaInvestigationId:@""
                          identifications:@[]
                          quantifications:@[]
                        provenanceRecords:@[]
                                    width:w
                                   height:h
                           spectralPoints:s
                                 tileSize:32
                               pixelSizeX:1.0
                               pixelSizeY:1.0
                              scanPattern:@"raster"
                                     cube:cube
                                   mzAxis:mz];

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w2 =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w2 writeStreamHeaderWithFormatVersion:@"1.2"
                                           title:@"img-mixed"
                                isaInvestigation:@""
                                        features:@[TTIOTransportV011Feature]
                                       nDatasets:0
                                           error:&err],
         "5.1 mixed: StreamHeader");
    PASS([w2 writeImageProcessed:img error:&err],
         "5.1 mixed: writeImageProcessed");
    PASS([w2 writeEndOfStreamWithError:&err], "5.1 mixed: EndOfStream");

    NSString *outPath = makeTempPathIp(@"mixed");
    unlink([outPath fileSystemRepresentation]);
    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    PASS([r writeTtioToPath:outPath error:&err] && err == nil,
         "5.1 mixed: materialised");
    TTIOMSImage *back = [TTIOMSImage readFromFilePath:outPath error:&err];
    PASS(back != nil, "5.1 mixed: re-opened");
    PASS([back.cube isEqualToData:img.cube],
         "5.1 mixed: round-trip cube byte-equal (mixed sparsity)");
    unlink([outPath fileSystemRepresentation]);
}

// -------- entry point ------------------------------------------------------

void testTransportImageProcessed(void);
void testTransportImageProcessed(void)
{
    testProcessedModeRoundTripsSparseCube();
    testAllZeroPixelEmitsNonzeroCountZero();
    testFullyDensePixelRoundTrips();
    testMixedSparseAndDensePixels();
}
